### PR TITLE
Fix issues found in image cache

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
### Description of the feature or fix

See the commit message.

### Checkpoints
- [X] Follow the [styling guide](https://github.com/lvgl/lvgl/blob/master/docs/CODING_STYLE.md)
- [ ] Update CHANGELOG.md
- [ ] Update the documentation

### CODE REVIEW

File: src/lv_draw/lv_img_cache.c [hunk 1]
```diff
@@ -33,13 +33,15 @@
...
-#if LV_IMG_CACHE_DEF_SIZE == 0
-    static lv_img_cache_entry_t cache_temp;
-#endif
+static bool lv_img_cache_match(const void * src1, const void * src2);
 
 /**********************
  *  STATIC VARIABLES
  **********************/
+#if LV_IMG_CACHE_DEF_SIZE == 0
+    static lv_img_cache_entry_t cache_temp;
```
Comment (amirgon) [inline src/lv_draw/lv_img_cache.c:41]: Shouldn't this be a GC root?

Comment (kisvegabor) [reply src/lv_draw/lv_img_cache.c:41]: As memory can be allocated inside `cache_temp` it really should be GC root. (Similarly to [_lv_img_cache_array](https://github.com/lvgl/lvgl/blob/351008edcc01753018ecc6a7a309fa01f2d42a81/src/lv_misc/lv_gc.h#L38))

But we can't use `#if` inside define so disabled cache config should be rather handled via a `_lv_img_cache_array[1]` array. I'll add this to the PR in a few minutes.

Comment (xiaoxiang781216) [reply src/lv_draw/lv_img_cache.c:41]: @kisvegabor I am investigating the solution too.

Comment (kisvegabor) [reply src/lv_draw/lv_img_cache.c:41]: Added in eb8bfb9

Comment (kisvegabor) [reply src/lv_draw/lv_img_cache.c:41]: @xiaoxiang781216 Let me know if you have  a better idea

Comment (xiaoxiang781216) [reply src/lv_draw/lv_img_cache.c:41]: @kisvegabor your change looks simpler than mine, so I incoperate your change with minor modification, please take a look.

**Comment by kisvegabor** [comment]
Looks good to me! 
@amirgon could you also take look to be sure it's "GC ROOT correct".

**Comment by amirgon** [comment]
> @amirgon could you also take look to be sure it's "GC ROOT correct".

Looks good, but only as long as `LV_GC_ROOT(_lv_img_cache_array)` is always assigned to `cache_single` when `cache_single` is compiled in.  
Currently the assignment is in `lv_img_cache_set_size` so the assumption is that `lv_img_cache_set_size` is in fact called. How do we ensure it's always called?

To make this clearer, why not always assign `LV_GC_ROOT(_lv_img_cache_array)` instead of assigning `cache_single`?

https://github.com/lvgl/lvgl/blob/cedbfe421a1ae3956fc810499d6c028f04d68cc1/src/lv_draw/lv_img_cache.c#L121-L123

**Comment by xiaoxiang781216** [comment]
> > @amirgon could you also take look to be sure it's "GC ROOT correct".
> 
> Looks good, but only as long as `LV_GC_ROOT(_lv_img_cache_array)` is always assigned to `cache_single` when `cache_single` is compiled in.
> Currently the assignment is in `lv_img_cache_set_size` so the assumption is that `lv_img_cache_set_size` is in fact called. How do we ensure it's always called?

This patch modify lv_init to ensure lv_img_cache_set_size get called at least once: 
https://github.com/lvgl/lvgl/pull/2081/files#diff-109d5ae74897190784b79014ca5ed5427f7969628b53d686a87cb066795fdb85L219-L221
> 
> To make this clearer, why not always assign `LV_GC_ROOT(_lv_img_cache_array)` instead of assigning `cache_single`?
> 
> https://github.com/lvgl/lvgl/blob/cedbfe421a1ae3956fc810499d6c028f04d68cc1/src/lv_draw/lv_img_cache.c#L121-L123

it's little bit quick to access `cache_single` directly instead through `LV_GC_ROOT(_lv_img_cache_array)`. If LV_IMG_CACHE_DEF_SIZE == 0, there is not difference between `LV_GC_ROOT(_lv_img_cache_array)` and `&cache_single`.

**Comment by amirgon** [comment]
> it's little bit quick to access `cache_single` directly instead through `LV_GC_ROOT(_lv_img_cache_array)`. If LV_IMG_CACHE_DEF_SIZE == 0, there is not difference between `LV_GC_ROOT(_lv_img_cache_array)` and `&cache_single`.

What's the advantage of defining `cache_single` as a static variable separate from gc_roots?  
Instead, you could define it directly on `LV_ITERATE_ROOTS` (not as a pointer).  
That would save the need to assign it to gc_roots on runtime.

Something like:
```c
#define LV_ITERATE_ROOTS(f) \
...
    f(lv_img_cache_entry_t, _lv_cache_single)                  \
```

**Comment by xiaoxiang781216** [comment]
> > it's little bit quick to access `cache_single` directly instead through `LV_GC_ROOT(_lv_img_cache_array)`. If LV_IMG_CACHE_DEF_SIZE == 0, there is not difference between `LV_GC_ROOT(_lv_img_cache_array)` and `&cache_single`.
> 
> What's the advantage of defining `cache_single` as a static variable separate from gc_roots?
> Instead, you could define it directly on `LV_ITERATE_ROOTS` (not as a pointer).
> That would save the need to assign it to gc_roots on runtime.
> 
> Something like:
> 
> ```c
> #define LV_ITERATE_ROOTS(f) \
> ...
>     f(lv_img_cache_entry_t, _lv_cache_single)                  \
> ```

Is it good to change the external public symbol definition base on the real config? I don't think so. If we follow your suggestion, LV_ITERATE_ROOTS need more change like this(pseudo code):
```
#define LV_ITERATE_ROOTS(f) \
    f(lv_ll_t, _lv_task_ll)  /*Linked list to store the lv_tasks*/ \
    f(lv_ll_t, _lv_disp_ll)  /*Linked list of screens*/            \
    f(lv_ll_t, _lv_indev_ll) /*Linked list of input device*/       \
#if LV_USE_FILESYSTEM
    f(lv_ll_t, _lv_drv_ll)                                         \
#endif
    f(lv_ll_t, _lv_file_ll)                                        \
#if LV_USE_ANIMATION
    f(lv_ll_t, _lv_anim_ll)                                        \
#endif
#if LV_USE_GROUP
    f(lv_ll_t, _lv_group_ll)                                       \
#endif
    f(lv_ll_t, _lv_img_defoder_ll)                                 \
    f(lv_ll_t, _lv_obj_style_trans_ll)                             \
#if LV_IMG_CACHE_DEF_SIZE
    f(lv_img_cache_entry_t*, _lv_img_cache_array)                  \
#else
    f(lv_img_cache_entry_t, _lv_img_cache)                         \
#endif
    f(lv_task_t*, _lv_task_act)                                    \
    f(lv_mem_buf_arr_t , _lv_mem_buf)                              \
    f(_lv_draw_mask_saved_arr_t , _lv_draw_mask_list)              \
#if LV_USE_THEME_MATERIAL
    f(void * , _lv_theme_material_styles)                          \
#endif
#if LV_USE_THEME_TEMPLATE
    f(void * , _lv_theme_template_styles)                          \
#endif
#if LV_USE_THEME_MONO
    f(void * , _lv_theme_mono_styles)                              \
#endif
#if LV_USE_THEME_EMPTY
    f(void * , _lv_theme_empty_styles)                             \
#endif
    f(uint8_t *, _lv_font_decompr_buf)                             \
```

**Comment by amirgon** [comment]
> Is it good to change the external public symbol definition base on the real config? I don't think so.

If they are not part of a public API, I don't see the problem.
Note the underscore at the beginning of each? These are not meant to be used in public context.
The only reason they are external symbols is because they are used in different compilation units of the same C module.

Does it make more sense defining a special static variable and paying cycles and ram just for assigning it elsewhere?

> If we follow your suggestion, LV_ITERATE_ROOTS need more change like this(pseudo code):

Actually I agree. It makes sense defining gc roots conditionally.  
But it should be done differently since in C we can't use `#if` inside macro definition.
Instead, we can do something like this:

```c
#define LV_ITERATE_ROOTS(f, f_cond) \
    f(lv_ll_t, _lv_task_ll)  /*Linked list to store the lv_tasks*/ \
    f(lv_ll_t, _lv_disp_ll)  /*Linked list of screens*/            \
    f(lv_ll_t, _lv_indev_ll) /*Linked list of input device*/       \
    f_cond(lv_ll_t, _lv_drv_ll, LV_USE_FILESYSTEM, 1)                                         \
...
    f_cond(lv_img_cache_entry_t*, _lv_img_cache_array, LV_IMG_CACHE_DEF_SIZE, 1)                  \
    f_cond(lv_img_cache_entry_t, _lv_img_cache, LV_IMG_CACHE_DEF_SIZE, 0)                         \
```

Then with just a little C preprocessor trickery, `f_cond` can be defined to be enabled/disabled according to the last two arguments, assuming these can be either 0 or 1.

**Comment by xiaoxiang781216** [comment]
Ok, I will provid a patch to clean up LV_ITERATE_ROOTS

**Comment by xiaoxiang781216** [comment]
@kisvegabor @amirgon I drop the gc related change from this PR, and will create a new PR to fix all gc related problem. So, I think this PR can be merged fist.

**Comment by xiaoxiang781216** [comment]
> > Is it good to change the external public symbol definition base on the real config? I don't think so.
> 
> If they are not part of a public API, I don't see the problem.
> Note the underscore at the beginning of each? These are not meant to be used in public context.
> The only reason they are external symbols is because they are used in different compilation units of the same C module.
> 
> Does it make more sense defining a special static variable and paying cycles and ram just for assigning it elsewhere?
> 
> > If we follow your suggestion, LV_ITERATE_ROOTS need more change like this(pseudo code):
> 
> Actually I agree. It makes sense defining gc roots conditionally.
> But it should be done differently since in C we can't use `#if` inside macro definition.
> Instead, we can do something like this:
> 
> ```c
> #define LV_ITERATE_ROOTS(f, f_cond) \
>     f(lv_ll_t, _lv_task_ll)  /*Linked list to store the lv_tasks*/ \
>     f(lv_ll_t, _lv_disp_ll)  /*Linked list of screens*/            \
>     f(lv_ll_t, _lv_indev_ll) /*Linked list of input device*/       \
>     f_cond(lv_ll_t, _lv_drv_ll, LV_USE_FILESYSTEM, 1)                                         \
> ...
>     f_cond(lv_img_cache_entry_t*, _lv_img_cache_array, LV_IMG_CACHE_DEF_SIZE, 1)                  \
>     f_cond(lv_img_cache_entry_t, _lv_img_cache, LV_IMG_CACHE_DEF_SIZE, 0)                         \
> ```
> 
> Then with just a little C preprocessor trickery, `f_cond` can be defined to be enabled/disabled according to the last two arguments, assuming these can be either 0 or 1.

@amirgon it's easy to define f_cond for macro which can be either 0 or 1, like below:
```
#define f_cond(f, t, n, m, v)   LV_CONCAT3(f_cond, m, v)(f, t, n)

#define f_cond00(f, t, n)       f(t, n)
#define f_cond01(f, t, n)
#define f_cond10(f, t, n)
#define f_cond11(f, t, n)       f(t, n)
```
But, LV_IMG_CACHE_DEF_SIZE can be any integer value, it's impossible to define f_cond to catch all possible combination.

**Comment by amirgon** [comment]
> But, LV_IMG_CACHE_DEF_SIZE can be any integer value, it's impossible to define f_cond to catch all possible combination.

Then define a helper macro like this:
```c
#if LV_IMG_CACHE_DEF_SIZE == 0
#define LV_IMG_CACHE_DEF 0
#else 
#define LV_IMG_CACHE_DEF 1
#endif
```

**Comment by kisvegabor** [comment]
Wouldn't it be the simplest if with `LV_IMG_CACHE_DEF_SIZE = 0` we allocated 1 entry into `_lv_img_cache_array` in `lv_img_cache_set_size`. In other words instead of using `cache_single` use a single dynamically allcoated entry.

**Comment by kisvegabor** [comment]
@xiaoxiang781216 Merged, thank you! Let's continue in the new PR for GC.

**Comment by xiaoxiang781216** [comment]
> Wouldn't it be the simplest if with `LV_IMG_CACHE_DEF_SIZE = 0` we allocated 1 entry into `_lv_img_cache_array` in `lv_img_cache_set_size`. In other words instead of using `cache_single` use a single dynamically allcoated entry.

This is actually my initial change. The most efficient way is to change/remove the global fields based config as @amirgon suggestion, but the code become a little bit complex

**Pull request merged**